File: mainframe/data/comp3.py
#!/usr/bin/env python3
"""COMP-3 packed decimal, as the mainframe writes it.

Packed decimal stores one decimal digit per 4-bit nibble, two digits per byte, with the final nibble
holding the sign. A signed field of n digits therefore occupies ceil((n + 1) / 2) bytes.

The estate's money field is ``PIC S9(13)V99 COMP-3``: fifteen digits, eight bytes. ``V`` is an implied
decimal point and occupies nothing, so the stored digits *are* the amount in minor units - for a
currency whose ISO 4217 scale is 2, which is the only kind stratum 0 can represent.

Packed decimal has **no endianness**. It is a digit string, not an integer, which is why a Java
encoder and a COBOL runtime produce identical bytes with no byte-order handling anywhere - and why
WP-11 can compare byte for byte rather than value for value.

See docs/architecture/canonical-data-model.md section 2 for the worked examples this implements.
"""

DIGITS = 15
"""Digits in PIC S9(13)V99: thirteen before the implied point, two after."""

SIZE = (DIGITS + 1) // 2 + (DIGITS + 1) % 2
"""Bytes occupied: ceil((15 + 1) / 2) = 8."""

POSITIVE = 0x0C
"""The only positive sign this estate writes. Zero uses it too."""

NEGATIVE = 0x0D
"""The only negative sign this estate writes."""

UNSIGNED = 0x0F
"""Written by fields declared without S. Accepted on read, never written."""

MAX = 10**DIGITS - 1


def encode_comp3(amount_minor: int, digits: int = DIGITS) -> bytes:
    """Pack ``amount_minor`` into COMP-3.

    Zero is always positive. A negative zero would break byte-for-byte comparison against the Java
    encoder in WP-11 and produce a reconciliation break in WP-16 that no posting caused.
    """
    limit = 10**digits - 1
    if not -limit <= amount_minor <= limit:
        raise ValueError(
            f"{amount_minor} does not fit in PIC S9({digits - 2})V99 COMP-3;"
            f" the range is -{limit} to {limit} minor units"
        )

    sign = NEGATIVE if amount_minor < 0 else POSITIVE
    packed = str(abs(amount_minor)).zfill(digits)

    nibbles = [int(character) for character in packed] + [sign]
    return bytes(
        (nibbles[i] << 4) | nibbles[i + 1] for i in range(0, len(nibbles), 2)
    )


def decode_comp3(raw: bytes, digits: int = DIGITS) -> int:
    """Unpack COMP-3 back to minor units.

    Accepts ``0x0F`` as positive because unsigned fields elsewhere in the estate write it, even
    though nothing here ever does.
    """
    expected = (digits + 1) // 2 + (digits + 1) % 2
    if len(raw) != expected:
        raise ValueError(f"expected {expected} bytes for {digits} digits, got {len(raw)}")

    nibbles = []
    for byte in raw:
        nibbles.append(byte >> 4)
        nibbles.append(byte & 0x0F)

    sign = nibbles.pop()
    if sign not in (POSITIVE, NEGATIVE, UNSIGNED):
        raise ValueError(f"unrecognised sign nibble 0x{sign:X}; expected C, D or F")

    value = int("".join(str(nibble) for nibble in nibbles))
    return -value if sign == NEGATIVE else value


def to_hex(raw: bytes) -> str:
    """A hex dump in the form the canonical data model prints, for error messages."""
    return " ".join(f"0x{byte:02X}" for byte in raw)

File: integration/esb-adapter/src/main/java/bank/tessera/esb/Comp3.java
package bank.tessera.esb;

/**
 * COMP-3 packed decimal, as the mainframe writes it.
 *
 * <p>The estate's money field is {@code PIC S9(13)V99 COMP-3}: fifteen digits in eight bytes,
 * {@code ceil((15 + 1) / 2)}. Packed decimal stores one decimal digit per 4-bit nibble, two per
 * byte, high nibble first, with the <strong>final nibble holding the sign</strong>. {@code V} is an
 * implied decimal point that occupies nothing, so the stored digits <em>are</em> {@code amountMinor}
 * - but only for a currency whose ISO 4217 scale is 2, which is what
 * {@link CurrencyScales#requireRepresentableOnTheMainframe(String)} is for.
 *
 * <p>Packed decimal has <strong>no endianness</strong>. It is a digit string, not an integer, so
 * this encoder and a COBOL runtime produce identical bytes with no byte-order handling anywhere -
 * which is the whole reason {@code Comp3Test} can compare byte for byte rather than value for value.
 *
 * <p>Two things here are easy to get wrong and produce a file every reader accepts:
 *
 * <ul>
 * <li><strong>Zero is positive.</strong> A {@code -0} would break byte-for-byte comparison against
 * the mainframe's own fixtures and manufacture a reconciliation break in WP-16 that no posting
 * caused.</li>
 * <li><strong>The last byte is not sign-only.</strong> It carries the fifteenth digit in its high
 * nibble: {@code +1} is {@code ...0x1C}, not {@code ...0x0C}.</li>
 * </ul>
 *
 * <p>{@code 0x0F} - unsigned - is what a field declared without {@code S} writes. It is accepted on
 * read elsewhere in the estate and is never written here; an encoder that emitted it would produce a
 * file the mainframe reads as unsigned, silently.
 *
 * <p>The reference implementation is {@code mainframe/data/comp3.py} and the arbiter is the worked
 * examples in {@code docs/architecture/canonical-data-model.md} section 2.
 */
public final class Comp3 {

    /** Digits in {@code PIC S9(13)V99}: thirteen before the implied point, two after. */
    public static final int DIGITS = 15;

    /** Bytes occupied: {@code ceil((15 + 1) / 2)}. */
    public static final int SIZE = 8;

    /** The largest amount the picture clause can hold, in minor units. */
    public static final long MAX = 999_999_999_999_999L;

    /** The only positive sign this estate writes. Zero uses it too. */
    private static final int POSITIVE = 0x0C;

    /** The only negative sign this estate writes. */
    private static final int NEGATIVE = 0x0D;

    private Comp3() {
    }

    /**
     * Pack an amount in minor units into eight bytes of COMP-3.
     *
     * @throws TransferHandlingException permanently, at stage {@code ENCODE}, when the amount does
     *     not fit. Refusal rather than truncation: a silently narrowed amount is the defect that
     *     prints as an ordinary-looking number, and re-presenting the same message would produce the
     *     same overflow, so retrying it would achieve nothing.
     */
    public static byte[] encode(long amountMinor) {
        if (amountMinor > MAX || amountMinor < -MAX) {
            throw TransferHandlingException.permanent(FailureStage.ENCODE,
                    "amount " + amountMinor + " minor units does not fit in PIC S9(13)V99 COMP-3;"
                            + " the range is -" + MAX + " to " + MAX);
        }

        // Left-zero-padded to the full fifteen digits, then read one nibble at a time. Math.abs is
        // safe here only because the range check above already excluded Long.MIN_VALUE.
        char[] digits = pad(Long.toString(Math.abs(amountMinor)));

        byte[] packed = new byte[SIZE];
        for (int i = 0; i < SIZE - 1; i++) {
            packed[i] = (byte) ((value(digits[i * 2]) << 4) | value(digits[i * 2 + 1]));
        }
        // The last byte pairs the fifteenth digit with the sign, never the sign alone.
        int sign = amountMinor < 0 ? NEGATIVE : POSITIVE;
        packed[SIZE - 1] = (byte) ((value(digits[DIGITS - 1]) << 4) | sign);
        return packed;
    }

    private static char[] pad(String digits) {
        char[] padded = new char[DIGITS];
        int leading = DIGITS - digits.length();
        for (int i = 0; i < leading; i++) {
            padded[i] = '0';
        }
        digits.getChars(0, digits.length(), padded, leading);
        return padded;
    }

    private static int value(char digit) {
        return digit - '0';
    }
}
